polynote-kernel/src/main/scala/polynote/kernel/util/NotebookContext.scala

```diff
@@ -0,0 +1,141 @@
+package polynote.kernel.util
+
+import cats.effect.{Concurrent, IO}
+import polynote.kernel.ResultValue
+import polynote.messages.{CellID, Notebook}
+
+import scala.annotation.tailrec
+
+/**
+  * Tracks all the CellContexts for a notebook. A good old-fashioned linked list in reverse, with an extra reference to the
+  * tip for easy append (i.e. predefs)
+  */
+class NotebookContext(implicit concurrent: Concurrent[IO]) {
+
```

Comment:
Lots of icky mutable stuff in these classes, but I feel fine about it 😬 I started with `Ref`s everywhere instead, but it was just like, why?

Reply:
I think one (of the many) issues with my previous attempt to work on this issue was trying to limit myself to "pure" code which lead to unnecessary complications. Someone wise recently told me not to over-FP-ify things too much and I think that's absolutely correct 😉